src/main/java/org/swisspush/reststorage/FileSystemDirLister.java

```diff
@@ -0,0 +1,137 @@
+                    // Set fields according to documentation in Resource class.
+                    name = Paths.get(path).getFileName().toString();
+                    exists = false;
+                    error = rejected = invalid = true;
+                    errorMessage = invalidMessage = event.cause().getMessage();
+                }};
+                handler.handle(erroneousResource);
+            } else {
+                handler.handle((Resource) event.result());
+            }
+        });
+        log.trace("Work delegated.");
+    }
+
+    private void listDirBlocking(String path, int offset, int count, Future<CollectionResource> future) {
+        //
+        // HINT: This method gets executed on a worker thread!
+        //
+        // Convert String to Path
+        final Path searchPath = Paths.get(canonicalizeVirtualPath(path));
+        // Prepare our result.
+        final CollectionResource collection = new CollectionResource() {{
+            items = new ArrayList<>(128);
+        }};
+        final String fullPath = canonicalizeVirtualPath(path);
+        try (Stream<Path> source = Files.list(searchPath)) {
+            source.forEach(entry -> {
+                final String entryName = entry.getFileName().toString();
+                log.trace("Processing entry '{}'", entryName);
+                if (".tmp".equals(entryName) && fullPath.length() == root.length()) {
+                    // Ignore hidden '/.tmp/' directory.
+                    return;
+                }
+                // Create resource representing currently processed directory entry.
+                final Resource resource;
+                if (Files.isDirectory(entry)) {
+                    resource = new CollectionResource();
+                } else if (Files.isRegularFile(entry)) {
+                    resource = new DocumentResource();
+                } else {
+                    resource = new Resource();
+                    resource.exists = false;
+                }
+                resource.name = entryName;
+                collection.items.add(resource);
+            });
+        } catch (IOException e) {
+            future.fail(e);
+            return;
+        }
+        Collections.sort(collection.items);
+        // Don't know exactly what we do here now. Seems we check 'limit' for a range request.
+        int n = count;
+        if (n == -1) {
+            n = collection.items.size();
+        }
+        // Don't know exactly what we do here. But it seems we evaluate 'start' of a range request.
+        if (offset > -1) {
+            if (offset >= collection.items.size() || (offset + n) >= collection.items.size() || (offset == 0 && n == -1)) {
+                future.complete(collection);
+            } else {
+                collection.items = collection.items.subList(offset, offset + n);
+                future.complete(collection);
+            }
+        } else {
+            // TODO: Resolve future
```

Comment:
We have to find out first in which case offset can be <= -1
I think it does not happen and is some smelly remaining.

Reply:
Sounds to me like a candidate for some 400 response :)

Reply:
The current implementation never respondsif we are in such a case. That means it is a bogus branch.
As no one complains about stuck request, I guess this branch (offset <= -1) is never used, thus the test is superfluous.
We should clear this up and remove the test so that we do not have weird dead code remains.

Reply:
Actually, offset is never negative. See: https://github.com/swisspush/vertx-rest-storage/blob/master/src/main/java/org/swisspush/reststorage/UrlParser.java#L15
Thus you can safely remove this test.